tests/msd/test_msd.py

```
@@ -0,0 +1,121 @@
+        return mask
+
+    def _fake_eval_abs(subgroup_mask, y_vec):
+        assert isinstance(y_vec, np.ndarray)
+        assert y_vec.ndim == 1
+        called["y"] = deepcopy(y_vec)
+        called["mask"] = deepcopy(subgroup_mask)
+        return 0.25
+
+    monkeypatch.setattr(msd, "subgroup_map_from_conjuncts_dataframe", _fake_subgroup_map, raising=True)
+    monkeypatch.setattr(msd, "evaluate_subgroup_discrepancy", _fake_eval_abs, raising=True)
+
+    val = msd.evaluate_MSD(X=X, y=y, rule=[("anything", "goes")], signed=False)
+    assert val == pytest.approx(0.25)
+    assert np.array_equal(called["mask"], mask)
+    assert np.array_equal(called["y"], np.array([1, 0, 1, 0]))
+
+def test_evaluate_MSD_signed(monkeypatch):
+    X = pd.DataFrame({"A": [0, 1, 0, 1]})
+    y = pd.Series([0, 1, 1, 0], dtype=int)
+
+    mask = np.array([False, True, False, True], dtype=bool)
+    seen = {}
+
+    def _fake_subgroup_map(rule, X_df):
+        return mask
+
+    def _fake_signed(subgroup_mask, y_vec):
+        # record inputs; return a distinct value to ensure correct branch used
+        seen["ndim"] = y_vec.ndim
+        return -0.123
+
+    monkeypatch.setattr(msd, "subgroup_map_from_conjuncts_dataframe", _fake_subgroup_map, raising=True)
+    monkeypatch.setattr(msd, "signed_subgroup_discrepancy", _fake_signed, raising=True)
+
+    val = msd.evaluate_MSD(X=X, y=y, rule=[("r", "u")], signed=True)
+    assert val == pytest.approx(-0.123)
+    assert seen["ndim"] == 1  # flattened/1D
+
+def test_evaluate_MSD_returns_float(monkeypatch):
+    monkeypatch.setattr(
+        msd, "subgroup_map_from_conjuncts_dataframe",
+        lambda rule, X_df: np.array([True, False], dtype=bool),
+        raising=True,
+    )
+    monkeypatch.setattr(
+        msd, "evaluate_subgroup_discrepancy",
+        lambda mask, y_vec: 0.0,
+        raising=True,
+    )
+    out = msd.evaluate_MSD(
+        X=pd.DataFrame({"A": [0, 1]}),
+        y=np.array([0, 1]),
+        rule=[("x", "y")],
```

this should raise an error as the rule is not in the correct format. I see that the function "map from conjuncts" is replaced here, but maybe something to add in general? Also I've noticed that the "map from conjuncts" function assumes that rule has a slightly different type than this function, that should also be corrected to list tuple (int, bin) I think.